### PR TITLE
Don't pre-allocate positions on handler for create/alter table

### PR DESCRIPTION
**Pull request by mfussenegger**
There can always be a race between analysing the statement on the
coordinator and processing the cluster state update. We "fixed" the
position in the cluster state update to account for that.

This changes the approach slightly. The handler doesn't attempt to
pre-allocate a position but instead we rely on the order of the columns
and the cluster state update increments the position for each _new_
column.

### CODE REVIEW

File: server/src/test/java/io/crate/integrationtests/CreateTableIntegrationTest.java [hunk 1]
```diff
@@ -194,22 +195,21 @@ tn boolean as (ta = false)
...
-             ti['tk']['tl']| 12
-             tm| 13
-             tn| 14
-             """); // 'th' is a named index and is assigned column position 8
+        assertThat(response).hasRows(
+             "ta| 1",
+             "tb| 2",
+             "tc| 3",
+             "tc['td']| 4",
+             "tc['te']| 5",
```
Comment (mfussenegger) [inline server/src/test/java/io/crate/integrationtests/CreateTableIntegrationTest.java:203]: The new approach changes the way nested columns are incremented.

I think that should be fine and it still fits the "in definition" order criteria.

File: server/src/test/java/io/crate/integrationtests/DynamicMappingUpdateITest.java [hunk 2]
```diff
@@ -23,6 +23,7 @@
 
 import static io.crate.testing.Asserts.assertThat;
 import static io.crate.testing.TestingHelpers.printedTable;
+import static org.assertj.core.api.Assertions.assertThat;
```
Comment (matriv) [inline server/src/test/java/io/crate/integrationtests/DynamicMappingUpdateITest.java:26]: not needed?

File: server/src/test/java/io/crate/integrationtests/CreateTableIntegrationTest.java (hunk 1 above)
Comment (matriv) [reply server/src/test/java/io/crate/integrationtests/CreateTableIntegrationTest.java:203]: but for this test, there is no change, or am I not seeing it?

File: server/src/main/java/io/crate/metadata/SimpleReference.java [hunk 3]
```diff
@@ -221,11 +221,11 @@ public boolean isGenerated() {
     }
 
     @Override
-    public Map<String, Object> toMapping() {
+    public Map<String, Object> toMapping(int position) {
         DataType<?> innerType = ArrayType.unnest(type);
         Map<String, Object> mapping = new HashMap<>();
         mapping.put("type", DataTypes.esMappingNameFrom(innerType.id()));
-        mapping.put("position", position);
+        mapping.put("position", position >= 0 ? position : this.position);
```
Comment (matriv) [inline server/src/main/java/io/crate/metadata/SimpleReference.java:228]: Apologies, maybe I'm confused, couldn't just use the `position`  arg and completely ignore the `this.position`? As far as I see the `position` arg is correctly calculated in `MappingUtil#addColumnProperties()`.

**Review by matriv — COMMENTED** [review]
I've left some questions, one more:
I don't get how the race condition is avoided? incrementing takes place in the new `AllocPosition`, so if 2 threads (or 2 different nodes) attempt to make some change concurrently, the will get the max position from `DocTableInfo` and each one increment it by 1. Am I missing something in the logic?

File: server/src/test/java/io/crate/integrationtests/CreateTableIntegrationTest.java (hunk 1 above)
Comment (mfussenegger) [reply server/src/test/java/io/crate/integrationtests/CreateTableIntegrationTest.java:203]: Right, I commented it on the wrong one, it's visible in the others.

File: server/src/test/java/io/crate/integrationtests/DynamicMappingUpdateITest.java (hunk 2 above)
Comment (mfussenegger) [reply server/src/test/java/io/crate/integrationtests/DynamicMappingUpdateITest.java:26]: For some reason eclipse adds it back automatically if I remove it.

**Comment by mfussenegger** [comment]
> I don't get how the race condition is avoided?

It's all run within a cluster state update task. The important bit is that the `DocTableInfo` instance that's used is based on the a `ClusterState` instance from the ClusterStateTaskExecutor `execute`, to ensure that it is the latest state, and that there is no other concurrent modifications happening.

Before the call to `TransportSchemaUpdateAction.populateColumnPositions(indexMapping);` did something similar (except on the Map<String, Object> properties, which is more complex as it needs to contain the traversal logic as well)

File: server/src/main/java/io/crate/metadata/SimpleReference.java (hunk 3 above)
Comment (mfussenegger) [reply server/src/main/java/io/crate/metadata/SimpleReference.java]: This was mostly for the tests. I added a fixup to change it.

Comment (matriv) [reply server/src/main/java/io/crate/metadata/SimpleReference.java]: Thx, I think it's more clear now in terms of prod code.

**Review by matriv — APPROVED** [review]
LGTM. Thank you!

**Pull request merged**